# Домашнее задание 1
Анализ датасета с помощью Pandas
🌟 Скачать данные файл kc_house_data.csv (Дополнительный
материал)
🌟 Считать данные с помощью pandas
🌟 Вывести на экран первые 5 строк
🌟 Посмотреть на описание признаков и на их содержание

In [1]:
import pandas as pd

pd.set_option('display.max_colwidth', None)
pd.set_option('display.precision', 6)
pd.set_option('display.float_format', '{:,.2f}'.format)

### Считать данные с помощью pandas

In [2]:
hd = pd.read_csv('./kc_house_data.csv')

### Вывести на экран первые 5 строк

In [3]:
hd.head()

id             date      price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000 221,900.00         3       1.00         1180   
1  6414100192  20141209T000000 538,000.00         3       2.25         2570   
2  5631500400  20150225T000000 180,000.00         2       1.00          770   
3  2487200875  20141209T000000 604,000.00         4       3.00         1960   
4  1954400510  20150218T000000 510,000.00         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650    1.00           0     0  ...      7        1180              0   
1      7242    2.00           0     0  ...      7        2170            400   
2     10000    1.00           0     0  ...      6         770              0   
3      5000    1.00           0     0  ...      7        1050            910   
4      8080    1.00           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode   lat    long  sqft_living15  sqft_lot15  
0      1955             0    98178 47.51 -122.26           1340        5650  
1      1951          1991    98125 47.72 -122.32           1690        7639  
2      1933             0    98028 47.74 -122.23           2720        8062  
3      1965             0    98136 47.52 -122.39           1360        5000  
4      1987             0    98074 47.62 -122.05           1800        7503  

[5 rows x 21 columns]

### Посмотреть на описание признаков и на их содержание
Судя по описанию данный файл является датасетом по недвижимости в городе Сиэтл, штат Вашингтон, США.

## Добавим новые столбцы
Изменим столбец date с типом object (строка) на столбец sale_date с типом datetime64

In [4]:
hd['sale_date'] = hd['date'].astype("datetime64[ns]")
hd = hd.drop('date', axis=1)

Изменим столбец waterfront с типом данных int64 на столбец с булевыми значениями

In [5]:
hd['is_waterfront'] = hd['waterfront'].astype(bool)
hd = hd.drop('waterfront', axis=1)
hd.head()

id      price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
0  7129300520 221,900.00         3       1.00         1180      5650    1.00   
1  6414100192 538,000.00         3       2.25         2570      7242    2.00   
2  5631500400 180,000.00         2       1.00          770     10000    1.00   
3  2487200875 604,000.00         4       3.00         1960      5000    1.00   
4  1954400510 510,000.00         3       2.00         1680      8080    1.00   

   view  condition  grade  ...  sqft_basement  yr_built  yr_renovated  \
0     0          3      7  ...              0      1955             0   
1     0          3      7  ...            400      1951          1991   
2     0          3      6  ...              0      1933             0   
3     0          5      7  ...            910      1965             0   
4     0          3      8  ...              0      1987             0   

   zipcode   lat    long  sqft_living15  sqft_lot15  sale_date is_waterfront  
0    98178 47.51 -122.26           1340        5650 2014-10-13         False  
1    98125 47.72 -122.32           1690        7639 2014-12-09         False  
2    98028 47.74 -122.23           2720        8062 2015-02-25         False  
3    98136 47.52 -122.39           1360        5000 2014-12-09         False  
4    98074 47.62 -122.05           1800        7503 2015-02-18         False  

[5 rows x 21 columns]

# Домашнее задание 2
## Проведите первичный анализ данных


1. Изучите типы данных


In [6]:
hd.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   id             21613 non-null  int64         
 1   price          21613 non-null  float64       
 2   bedrooms       21613 non-null  int64         
 3   bathrooms      21613 non-null  float64       
 4   sqft_living    21613 non-null  int64         
 5   sqft_lot       21613 non-null  int64         
 6   floors         21613 non-null  float64       
 7   view           21613 non-null  int64         
 8   condition      21613 non-null  int64         
 9   grade          21613 non-null  int64         
 10  sqft_above     21613 non-null  int64         
 11  sqft_basement  21613 non-null  int64         
 12  yr_built       21613 non-null  int64         
 13  yr_renovated   21613 non-null  int64         
 14  zipcode        21613 non-null  int64         
 15  lat            2161

2. Найдите количество пропущенных ячеек в данных


In [7]:
hd.isna().sum()

id               0
price            0
bedrooms         0
bathrooms        0
sqft_living      0
sqft_lot         0
floors           0
view             0
condition        0
grade            0
sqft_above       0
sqft_basement    0
yr_built         0
yr_renovated     0
zipcode          0
lat              0
long             0
sqft_living15    0
sqft_lot15       0
sale_date        0
is_waterfront    0
dtype: int64

In [8]:
hd.isnull().sum()

id               0
price            0
bedrooms         0
bathrooms        0
sqft_living      0
sqft_lot         0
floors           0
view             0
condition        0
grade            0
sqft_above       0
sqft_basement    0
yr_built         0
yr_renovated     0
zipcode          0
lat              0
long             0
sqft_living15    0
sqft_lot15       0
sale_date        0
is_waterfront    0
dtype: int64

Пропущенных ячеек нет.

3. Посчитайте основные статистики по всем признакам и поизучайте их

In [9]:

hd.describe()

id        price  bedrooms  bathrooms  sqft_living  \
count        21,613.00    21,613.00 21,613.00  21,613.00    21,613.00   
mean  4,580,301,520.86   540,088.14      3.37       2.11     2,079.90   
min       1,000,102.00    75,000.00      0.00       0.00       290.00   
25%   2,123,049,194.00   321,950.00      3.00       1.75     1,427.00   
50%   3,904,930,410.00   450,000.00      3.00       2.25     1,910.00   
75%   7,308,900,445.00   645,000.00      4.00       2.50     2,550.00   
max   9,900,000,190.00 7,700,000.00     33.00       8.00    13,540.00   
std   2,876,565,571.31   367,127.20      0.93       0.77       918.44   

          sqft_lot    floors      view  condition     grade  sqft_above  \
count    21,613.00 21,613.00 21,613.00  21,613.00 21,613.00   21,613.00   
mean     15,106.97      1.49      0.23       3.41      7.66    1,788.39   
min         520.00      1.00      0.00       1.00      1.00      290.00   
25%       5,040.00      1.00      0.00       3.00      7.00    1,190.00   
50%       7,618.00      1.50      0.00       3.00      7.00    1,560.00   
75%      10,688.00      2.00      0.00       4.00      8.00    2,210.00   
max   1,651,359.00      3.50      4.00       5.00     13.00    9,410.00   
std      41,420.51      0.54      0.77       0.65      1.18      828.09   

       sqft_basement  yr_built  yr_renovated   zipcode       lat      long  \
count      21,613.00 21,613.00     21,613.00 21,613.00 21,613.00 21,613.00   
mean          291.51  1,971.01         84.40 98,077.94     47.56   -122.21   
min             0.00  1,900.00          0.00 98,001.00     47.16   -122.52   
25%             0.00  1,951.00          0.00 98,033.00     47.47   -122.33   
50%             0.00  1,975.00          0.00 98,065.00     47.57   -122.23   
75%           560.00  1,997.00          0.00 98,118.00     47.68   -122.12   
max         4,820.00  2,015.00      2,015.00 98,199.00     47.78   -121.31   
std           442.58     29.37        401.68     53.51      0.14      0.14   

       sqft_living15  sqft_lot15                      sale_date  
count      21,613.00   21,613.00                          21613  
mean        1,986.55   12,768.46  2014-10-29 04:38:01.959931648  
min           399.00      651.00            2014-05-02 00:00:00  
25%         1,490.00    5,100.00            2014-07-22 00:00:00  
50%         1,840.00    7,620.00            2014-10-16 00:00:00  
75%         2,360.00   10,083.00            2015-02-17 00:00:00  
max         6,210.00  871,200.00            2015-05-27 00:00:00  
std           685.39   27,304.18                            NaN

4. Пишите выводы

Данный файл является датасетом недвижимости на 61 авеню, города Сиэтл, штат Вашингтон, США.
Содержит информацию о ценах, площади, качестве, количестве комнат и об окружении 21613 объектах недвижимости. 

# Домашнее задание 3
## Ответьте на несколько вопросов

3.1 В каком диапазоне изменяются стоимости недвижимости?

In [10]:
min_price = hd['price'].min()
max_price = hd['price'].max()
f'Стоимость недвижимости изменяется в диапазоне от {min_price:,.2f} до {max_price:,.2f}'

'Стоимость недвижимости изменяется в диапазоне от 75,000.00 до 7,700,000.00'

3.2 Какую долю в среднем занимают жилая площадь от всей площади по всем домам?

In [11]:
total_living_sqft = hd['sqft_living'].sum()
total_lot_sqft = hd['sqft_lot'].sum()
total_objects = hd['id'].count()
print(
    f"Доля занимаемая жилой площадью ({hd['sqft_living'].sum():.0f}) от общей площади дома ({hd['sqft_lot'].sum():.0f}) в среднем составляет {hd['sqft_living'].mean() / hd['sqft_lot'].mean() * 100:.2f}%")

Доля занимаемая жилой площадью (44952873) от общей площади дома (326506890) в среднем составляет 13.77%


3.3 Как много домов с разными этажами в данных?

In [12]:
print("Домов с разными этажами:\n")
hd['floors'].value_counts()

Домов с разными этажами:


floors
1.00    10680
2.00     8241
1.50     1910
3.00      613
2.50      161
3.50        8
Name: count, dtype: int64

3.4 Насколько хорошие состояния у домов в данных?

In [13]:
hd['condition'].value_counts()

condition
3    14031
4     5679
5     1701
2      172
1       30
Name: count, dtype: int64

3.5 Найдите года, когда построили первый дом, когда построили последний дом в данных?

In [14]:
print(f"Первый дом построили в {hd['yr_built'].min()} году, последний дом построили в {hd['yr_built'].max()} году")
print(
    f"Первый дом в датасете построили в {hd.iloc[0]['yr_built']} году, последний дом в датасете построили в {hd.iloc[-1]['yr_built']} году")

Первый дом построили в 1900 году, последний дом построили в 2015 году
Первый дом в датасете построили в 1955 году, последний дом в датасете построили в 2008 году


# Домашнее задание 4
## Ответьте на несколько вопросов

4.1 Сколько в среднем стоят дома, у которых 2 спальни?

In [15]:
print(f"Дома у которых 2 спальни в среднем стоят {hd[hd['bedrooms'] == 2]['price'].mean():,.2f}")
two_bedrooms_house = hd[hd['bedrooms'] == 2]
total_price_2br_house = two_bedrooms_house['price'].sum()
print(
    f"Проверка: всего домов с двумя спальнями {two_bedrooms_house['id'].count()} шт. Общая стоимость этих домов {total_price_2br_house:,.2f}. Средняя стоимость равна {total_price_2br_house / two_bedrooms_house['id'].count():,.2f}")

Дома у которых 2 спальни в среднем стоят 401,372.68
Проверка: всего домов с двумя спальнями 2760 шт. Общая стоимость этих домов 1,107,788,602.00. Средняя стоимость равна 401,372.68


4.2 Какая в среднем общая площадь домов, у которых стоимость больше 600 000?

In [16]:
f"Общая площадь домов стоимость которых больше 600 000 в среднем равна {hd[hd['price'] > 600_000]['sqft_lot'].mean():,.2f}"

'Общая площадь домов стоимость которых больше 600 000 в среднем равна 20,442.52'

4.3 Как много домов коснулся ремонт?

In [17]:
f"{hd[hd['yr_renovated'] != 0]['id'].count()} домов были отремонтированы"

'914 домов были отремонтированы'

4.4 Насколько в среднем стоимость домов с оценкой grade домов выше 10 отличается от стоимости домов с оценкой grade меньше 4?

In [18]:
mean_price_grade_gr_10 = hd[hd['grade'] > 10]['price'].mean()
mean_price_grade_lt_4 = hd[hd['grade'] < 4]['price'].mean()
diff_price = mean_price_grade_gr_10 - mean_price_grade_lt_4
f"В среднем стоимость домов с оценкой grade выше 10 отличается от стоимости домов с оценкой grade меньше 4 на {diff_price:,.2f}"

'В среднем стоимость домов с оценкой grade выше 10 отличается от стоимости домов с оценкой grade меньше 4 на 1,488,885.12'

# Домашнее задание 5
## 5.1 Выберите дом клиенту
Клиент хочет дом с видом на набережную, как минимум с тремя
ванными и с подвалом. Сколько вариантов есть у клиента?

In [19]:
print(
    f"По данному запросу есть {hd[(hd['is_waterfront']) & (hd['bathrooms'] >= 3) & (hd['sqft_basement'] > 0)]['id'].count()} вариант(a,ов)")

hd[(hd['is_waterfront']) & (hd['bathrooms'] >= 3) & (hd['sqft_basement'] > 0)]

По данному запросу есть 41 вариант(a,ов)


id        price  bedrooms  bathrooms  sqft_living  sqft_lot  \
300    3225069065 3,075,000.00         4       5.00         4550     18641   
457    2122039094   705,000.00         3       3.00         1970     20978   
1164   1247600105 5,110,800.00         5       5.25         8010     45517   
1261   7636800041   995,000.00         3       4.50         4380     47044   
1315   7558700030 5,300,000.00         6       6.00         7390     24829   
1361   1925069082 2,200,000.00         5       4.25         4640     22703   
2626   7738500731 4,500,000.00         5       5.50         6640     40014   
2864   4114601570 3,600,000.00         3       3.25         5020     12431   
3914   9808700762 7,062,500.00         5       4.50        10040     37325   
3934   1269200229 1,379,900.00         3       3.25         3786     38038   
5760    623059016 1,100,000.00         4       3.25         3190     11774   
6233   2024069008 2,200,000.00         5       4.75         5990     10450   
6508   4217402115 3,650,000.00         6       4.75         5480     19401   
7989   9362000040 3,400,000.00         3       4.50         5230     17826   
8092   1924059029 4,668,000.00         5       6.75         9640     13068   
8163   1724069079 1,452,000.00         2       3.25         2070      3128   
8784   3343302110 1,800,000.00         3       3.00         2790     13295   
10263   625059036 2,700,000.00         5       4.00         4230     27295   
11257  3761100045 3,000,000.00         4       4.25         4850     12445   
11402  3426049284 2,300,000.00         4       3.25         4110     15929   
11424  5119400075   950,000.00         3       3.25         3050     18892   
11446   624069035 2,750,000.00         4       4.00         4130      5575   
11844  2623039082   770,000.00         3       3.50         2050     21744   
12459  4114601580 1,900,000.00         6       4.00         3020     13237   
12935  3222049044   835,000.00         3       3.00         2790     12523   
13710  2923500230 2,600,000.00         4       4.50         5270     12195   
13967  3126059023 3,395,000.00         4       3.50         4730     47870   
14083  3625059043 3,300,000.00         5       4.75         6200     13873   
14619  7936500109 2,230,000.00         3       3.00         3620     28064   
14822  3623500408 2,600,000.00         3       3.00         3410     16015   
15482   624069108 3,200,000.00         4       3.25         7000     28206   
15668  4030100290 1,680,000.00         5       3.50         5170      7197   
16817   825059178 2,574,000.00         4       3.75         4475     20424   
17767  7352200450 2,050,000.00         4       3.25         3580     19989   
18199  3625059152 3,300,000.00         3       3.25         4220     41300   
18455  8043700300 2,700,000.00         4       3.25         4420      7850   
18556  6613000930 2,950,000.00         4       3.25         3890     25470   
20325   518500480 3,000,000.00         3       3.50         4410     10756   
20767  8043700105 2,300,000.00         4       4.00         4360      8175   
21201   518500460 2,230,000.00         3       3.50         3760      5634   
21576  9253900271 3,567,000.00         5       4.50         4850     10584   

       floors  view  condition  grade  ...  sqft_basement  yr_built  \
300      1.00     4          3     10  ...           1950      2002   
457      2.00     3          4      9  ...            200      1980   
1164     2.00     4          3     12  ...           2020      1999   
1261     2.00     3          3      9  ...            660      1968   
1315     2.00     4          4     12  ...           2390      1991   
1361     2.00     4          5      8  ...           1780      1952   
2626     2.00     4          3     12  ...            290      2004   
2864     2.00     4          3     10  ...           1600      1941   
3914     2.00     2          3     11  ...           2360      1940   
3934     1.00     4         

## 5.2 Выберите дом клиенту
Клиент хочет дом либо с очень красивым видом из окна, либо с
видом на набережную, в очень хорошем состоянии и год
постройки не меньше 1980 года. В какой ценовом диапазоне
будут дома?


In [20]:
query = hd[((hd['is_waterfront']) | (hd['view'] >= 3)) & (hd['condition']>=4) & (hd['yr_built']>=1980)]
f"По данному запросу дома находятся в диапазоне цен от {query['price'].min():,.2f} до {query['price'].max():,.2f}"

'По данному запросу дома находятся в диапазоне цен от 420,000.00 до 5,300,000.00'

## 5.3 Выберите дом клиенту
Клиент хочет дом без подвала, с двумя этажами, стоимостью
до 150_000. Какая оценка по состоянию у таких домов в
среднем?

In [21]:
query = hd[(hd['sqft_basement']==0) & (hd['floors']==2) & (hd['price']<150_000)]
f"В среднем оценка по состоянию у домов из запроса равна {query['condition'].mean():.0f}"

'В среднем оценка по состоянию у домов из запроса равна 3'